File: topic/longest_common_subsequence/longest_common_subsequence.go
package longest_common_subsequence

func LongestCommonSubsequence(text1 string, text2 string) int {
	var last, tmp int
	dp := make([]int, len(text2))

	for i := 0; i < len(text1); i++ {
		last = dp[0]

		if text1[i] == text2[0] && last != 1 {
			last = 1
		}

		for j := 1; j < len(text2); j++ {
			if text1[i] == text2[j] {
				tmp = dp[j-1] + 1
			} else {
				tmp = max(last, dp[j])
			}
			dp[j-1] = last
			last = tmp
		}

		dp[len(text2)-1] = last
	}

	return dp[len(text2)-1]
}

func max(x, y int) int {
	if x > y {
		return x
	}
	return y
}

File: topic/longest_common_subsequence/py/longest_common_subsequence.py
class Solution:
    # DP solution after dimensionality reduction.
    # def longestCommonSubsequence(self, text1: str, text2: str) -> int:
    #     dp = [0] * len(text2)
    #     for _, char1 in enumerate(text1):
    #         pre = dp[0]
    #         dp[0] += 0 if char1 != text2[0] else 1
    #         for j, char2 in enumerate(text2[1:], start=1):
    #             tmp = dp[j]
    #             dp[j] = pre + 1 if char1 == char2 else max(dp[j-1], dp[j])
    #             pre = tmp
    #     return dp[-1]

    # DP solution.
    def longestCommonSubsequence2(self, text1: str, text2: str) -> int:
        dp = [[0] * (len(text2) + 1) for _ in range(len(text1) + 1)]
        for i, char1 in enumerate(text1):
            for j, char2 in enumerate(text2):
                dp[i + 1][j + 1] = 1 + \
                    dp[i][j] if char1 == char2 else max(
                        dp[i][j + 1], dp[i + 1][j])
        return dp[-1][-1]
